### ISSUE TITLE
bool custom Transformer for non zero-value false

### ISSUE THREAD

**Issue by mberlanda** [imdario/mergo#89]
hi @iamdario,

I recently introduced `mergo` on one of our projects.

I came up with an issue very similar to #24 :

```go
import (
  "testing"
  "github.com/imdario/mergo"
  "github.com/stretchr/testify/assert"
)
func TestBoolean(t *testing.T) {
  type Foo struct {
    Bar bool `json:"bar"`
  }

  f1 := Foo{Bar: true}
  f2 := Foo{Bar: false}

  mergo.Merge(&f2, f1)
  assert.False(t, f2.Bar)
}
```

Do you think it could be handled with some custom transformers?

**Comment by hectorj2f** [comment]
Is there any progress on this ? I am facing the same issue

**Comment by luisdavim** [comment]
I think this is the same issue reported in https://github.com/Praqma/helmsman/issues/137

**Comment by svyotov** [comment]
As far as I can see this is because of https://github.com/imdario/mergo/blob/9316a62528ac99aaecb4e47eadd6dc8aa6533d58/mergo.go#L36-L59
Merging Boolean and Numeric values that are false or zero are considered empty thus replaced.

**Comment by github4jiawen** [comment]
This can be accomplished with custom bool and transformer feature
```
package main

import (
	"encoding/json"
	"log"
	"reflect"

	"github.com/imdario/mergo"
)

type BoolInSetting struct {
	Set   bool
	Value bool
}

func (b *BoolInSetting) UnmarshalJSON(data []byte) error {
	b.Set = true

	err := json.Unmarshal(data, &b.Value)
	return err
}

type BoolInSettingTransformer struct {
}

func (t BoolInSettingTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
	if (typ == reflect.TypeOf(BoolInSetting{})) {
		return func(dst, src reflect.Value) error {
			if dst.CanSet() && src.FieldByName("Set").Bool() {
				dst.Set(src)
			}
			return nil
		}
	}
	return nil
}

type ConfigWithBool struct {
	BooleanField BoolInSetting `json:"BooleanField"`
}

func main() {
	var trusy ConfigWithBool
	var falsy ConfigWithBool
	var err error
	err = json.Unmarshal([]byte(`{"BooleanField": true}`), &trusy)
	if err != nil {
		log.Fatalln(err)
	}
	err = json.Unmarshal([]byte(`{"BooleanField": false}`), &falsy)
	if err != nil {
		log.Fatalln(err)
	}
	err = mergo.Merge(&trusy, falsy, mergo.WithTransformers(BoolInSettingTransformer{}))
	if err != nil {
		log.Fatalln(err)
	}
	log.Printf("%+v", trusy)
}

```

**Comment by vilipwong** [comment]
Hello, is `mergo.WithOverride` or `mergo.WithOverwriteWithEmptyValue` suppose to solve this ?

I faced the same issue with string and boolean value, using `mergo.WithOverwriteWithEmptyValue` e.g:
```golang
p1 := map[string]interface{}{
	"A": 3, "B": "note", "C": true,
}
p2 := map[string]interface{}{
	"B": "", "C": false,
}
mergo.Merge(&p1, &p2, mergo.WithOverwriteWithEmptyValue)
```

Expected result:
`{A:3 B:"" C:false}`

Actual result:
`{A:3 B:"note" C:true}`

**Comment by imdario** [comment]
@vilipwong Yes, `mergo.WithOverwriteWithEmptyValue` should help. I fixed this indirectly in #131. Anyway, I'm going to use your examples to make sure your use cases are covered.

**Comment by imdario** [comment]
Ok, let's see. The OP's example expects `f2.Bar` to be false, and it is the destination variable, so it's always truthy.

I rewrote the example with more meaningful variable names, as it was misleading me to the opposite direction:

```go
import (
  "testing"
  "github.com/imdario/mergo"
  "github.com/stretchr/testify/assert"
)
func TestBoolean(t *testing.T) {
  type Foo struct {
    Bar bool `json:"bar"`
  }

  f1 := Foo{Bar: true}
  f2 := Foo{Bar: false}

  mergo.Merge(&f2, f1)
  assert.False(t, f2.Bar)
}
```

Your intention, from what I understood, is to keep the false value as is. Not sure why as I'm missing context here. I'm going to tag along and restate what I said in #24:

> This is the expected behavior. As stated in README:
>> Mergo is intended to assign only zero value fields on destination with source value.

Provided this, your example won't work with the current implementation. If you need to protect false values, you must follow @github4jiawen example. Transformers are a simple way to filter the main behavior. 

About your case, @vilipwong, I pushed another test to check it out, and I confirm you that `mergo.WithOverwriteWithEmptyValue` will help you. Take into account that you have the opposite problem, as you want to override a truthy value with false.

**Comment by vilipwong** [comment]
I see. i was using release `v0.3.9`, it works when i go get `@master`. Any reason why its not released ? 
@imdario

**Comment by imdario** [comment]
@vilipwong I want to do the release with some more fixes from v0.3.9 related bugs.

**Comment by TomUnequals** [comment]
For those of you who need it, @Pothulapati built a fork that solves this [here](https://github.com/Pothulapati/mergo/commit/5a1b1cee7b3f9022825154dea4dcf5fb0dceecb8).
